### ISSUE TITLE
Question about 'label' in training of GAN

### ISSUE THREAD

**Issue by jihobak** [fchollet/deep-learning-with-python-notebooks#9]
Code below is part of '8.5-introduction-to-gans.ipynb' .
```
# Combine them with real images
stop = start + batch_size
real_images = x_train[start: stop]
combined_images = np.concatenate([generated_images, real_images])

# Assemble labels discriminating real from fake images
labels = np.concatenate([np.ones((batch_size, 1)),
                                          np.zeros((batch_size, 1))])
# Add random noise to the labels - important trick!
labels += 0.05 * np.random.random(labels.shape)

# Train the discriminator
d_loss = discriminator.train_on_batch(combined_images, labels)
```

The label of generated_image for training discriminator is '1' ( np.ones((batch_size, 1)).

As I understand GAN, The output of discriminator is the probability that the image is **real**.
It means label of generated_images for discriminator should be '0' because It is fake. However, Above code is not...

Thus, I think labels should be like below
labels = np.concatenate([np.zeros((batch_size, 1)),
                                          np.one((batch_size, 1))])

If this is wrong, Could you tell me why it is?

**Comment by jihobak** [comment]
Cross entropy is 

> z* -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x))

loss of discriminator which we try to minimize is 
> -(log(D(real)) + log(1-G(fake)))
> -log(D(real)) - log(1-G(fake))

That's why we should give '1' label to fake input in training Discriminator.
Now, I can understand your code.

My mistake :)

**Comment by USTClj** [comment]
Hi, I have a doubt in this ipynb, too. 
The doubt is that after we compile a model, say discriminator here, during defining the gan, dose 
'discriminator.trainable = False' affect the discriminator compiled before?

**Comment by hiroyachiba** [comment]
@jihobak, in the text description, real=1 and fake=0, but
in the program, real=0 and fake=1.
Although they should use the same convention, I think it works so long as the convention is consistent within the program (the following two places).

    # Assemble labels discriminating real from fake images
    labels = np.concatenate([np.ones((batch_size, 1)),
                             np.zeros((batch_size, 1))])

    # Assemble labels that say "all real images"
    misleading_targets = np.zeros((batch_size, 1))

**Comment by hiroyachiba** [comment]
@USTClj, I think so, too.  discriminator.trainable should be turned on/off at the following places.

    # Train the discriminator
    discriminator.trainable = True
    d_loss = discriminator.train_on_batch(combined_images, labels)

    # Train the generator (via the gan model,
    # where the discriminator weights are frozen)
    discriminator.trainable = False
    a_loss = gan.train_on_batch(random_latent_vectors, misleading_targets)

**Comment by hnomura** [comment]
I think trick here is, discriminator.trainable = False is after compiling discriminator (hence, discriminator is still trainable when used alone), and before compiling gan (which contains discriminator as a part of the network).  Therefore, when training gan, its containing discriminator is not trainable.
